### PR TITLE
Vectorize reviewer aggregation with bulk gather and dim=2 reduction

### LINKED ISSUE #385: Use matrix operations to select the scores for each reviewers
Once we have all the scores computed for all the submissions and all publications, use matrix operations to select the best scores for each reviewer considering:

- max, average or percentile
- publication weight

### PR DESCRIPTION
**Pull request by cagostino**
Closes #385.

Replaces the per-reviewer tensor slicing and reduction loop with a single bulk gather across all reviewers followed by a vectorized dim=2 reduction. 

Old code iterated reviewers one by one, with each iteration running a nested paper loop to build `train_paper_idx`, slicing `p2p_aff_norm[:, train_paper_idx]`, optionally weighting that 2D slice in logit space, reducing via `torch.quantile(..., dim=1)` or `.mean(dim=1)` or `.max(dim=1)[0]`, and appending to `score_vectors` for a final `torch.stack(..., dim=1)`. 

New code keeps the outer loop but strips it to just collecting valid paper indices via list comprehension. After the loop, a padding pass fills a dense `(num_reviewers, max_papers)` `paper_idx` tensor and boolean mask. 

Then `p2p_aff_norm[:, paper_idx]` gathers all reviewer-paper scores in one shot into `(num_test, num_reviewers, max_papers)`. 
Venue weights use `train_weight_tensor[paper_idx]` on the full 3D tensor. A single reduction across `dim=2` produces the final matrix: percentile via `torch.nanquantile` with NaN masking, average via zero-masked sum divided by counts, max via -inf masking. The PyTorch ops moved from once per reviewer inside the loop to once total outside the loop. Remaining loops only do Python list manipulation and tensor padding. Same change in specter.py and scincl.py.

### CODE REVIEW

**Comment by melisabok** [comment]
Chris, there are some conflicts here.

File: expertise/models/specter2_scincl/specter.py [hunk 1]
```diff
@@ -314,59 +314,54 @@ def load_emb_file(emb_file, paper_id_to_weight=None):
...
+            if reviewer_ids:
+                max_papers = max(len(papers) for papers in reviewer_paper_indices)
+                num_reviewers = len(reviewer_ids)
+                paper_idx = torch.zeros((num_reviewers, max_papers), dtype=torch.long)
+                mask = torch.zeros((num_reviewers, max_papers), dtype=torch.bool)
+                for r, papers in enumerate(reviewer_paper_indices):
+                    paper_idx[r, :len(papers)] = torch.tensor(papers, dtype=torch.long)
+                    mask[r, :len(papers)] = True
+
+                scores = p2p_aff_norm[:, paper_idx]  # (num_test, num_reviewers, max_papers)
```
[the remark this answers is not in the document]
Comment (cagostino) [reply expertise/models/specter2_scincl/specter.py:352]: We split machine type based on archive count before the job starts. For the actual cases we deal with, the 3D tensor fits without OOM.

For a sufficiently large job (~30k submissions, ~15k reviewers, max_papers ~40, float32):
- p2p_aff: ~76 GB
- scores_3d: ~72 GB
- weights_3d: ~72 GB (only if venue weights on)
- scores_matrix: ~2 GB
- Peak ~258 GB — fits within the memory on the tier those workloads are routed to.

The 3D tensor is temporary and freed immediately after reduction. If a future workload needs more headroom, we can batch along the submission dimension.

**Comment by melisabok** [comment]
as we discussed today, we should first commit a few tests verifying the output of the scores and the compare them with the changes of this PR. Am I missing something?

**Comment by cagostino** [comment]
> as we discussed today, we should first commit a few tests verifying the output of the scores and the compare them with the changes of this PR. Am I missing something?

working on translating those still into ci form, will re-request review when those are in and passing

**Comment by melisabok** [comment]
ok, I will this PR after the tests are in master

**Comment by carlosmondra** [comment]
Logically the changes look good to me. The operations described in the issue are addressed and improved with matrix operations. We need to test that the new 3D tensor operations fit in memory. They should, but @cagostino wants to try it out before merging. We are planning how to do it.

**Comment by cagostino** [comment]
in its current implementation, the aggregations do in fact produce OOM errors because the number of max papers is substantially higher than initially estimated. OOM was confirmed in a dev deployment and a run on a recent large prod job dataset. given the other improvements made in loop reductions, the gain from replacing these operations is not likely to be significant, this may be revisited after a proper embedding caching system is introduced and it presents itself as the next bottleneck.  closing for now
